Update Serialization For CoreIPCNSURLRequest

sheeparegreat opened the pull request:
#### b0bf239c98721c3235592515281e772c2f758dfe
<pre>
Update Serialization For CoreIPCNSURLRequest
<a href="https://bugs.webkit.org/show_bug.cgi?id=279092">https://bugs.webkit.org/show_bug.cgi?id=279092</a>
<a href="https://rdar.apple.com/134401345">rdar://134401345</a>

Reviewed by Chris Dumez.

Add NULL checks and update tests to run on more platforms.

* Source/WebKit/Shared/Cocoa/CoreIPCNSURLRequest.mm:
(WebKit::CoreIPCNSURLRequest::toID const):
* Tools/TestWebKitAPI/Tests/IPC/IPCSerialization.mm:
(TEST(IPCSerialization, NSURLRequest)):
(TEST(IPCSerialization, SecureCoding)):

Canonical link: <a href="https://commits.webkit.org/283236@main">https://commits.webkit.org/283236@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/d87849dc1b784f00ff957e8b09cca79387815b24

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/65612 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/44983 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/18230 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/69638 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/16221 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/67730 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/52784 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/16501 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/52684 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/11259 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/68679 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/41545 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/56792 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/33311 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/38224 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/14171 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/15097 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/60061 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/14531 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/71344 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/9566 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/13955 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/60002 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/9607 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/56857 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/60300 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/7912 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/1554 "Passed tests") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/9949 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/40793 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/41869 "Built successfully") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/43052 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/41613 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebKit/Shared/Cocoa/CoreIPCNSURLRequest.mm:
```diff
@@ -51,12 +51,17 @@ - (instancetype)_initWithWebKitPropertyListData:(NSDictionary *)plist;
+#define SET_DICT_FROM_OPTIONAL_MEMBER(KEY) \
+    if (m_data.KEY) { \
+        auto var##KEY = m_data.KEY->toID(); \
+        if (var##KEY) \
```
Could reduce the scope of the variable by using:
```
if (auto var##KEY = m_data.KEY->toID())
```

cdumez commented on Source/WebKit/Shared/Cocoa/CoreIPCNSURLRequest.mm:
```diff
@@ -51,12 +51,17 @@ - (instancetype)_initWithWebKitPropertyListData:(NSDictionary *)plist;
-    [dict setObject:m_data.KEY.toID().get() forKey:@#KEY]
+#define SET_DICT_FROM_MEMBER(KEY)                                   \
+    auto var##KEY = m_data.KEY.toID(); \
+    if (var##KEY) \
```
ditto

cdumez commented on Source/WebKit/Shared/Cocoa/CoreIPCNSURLRequest.mm:
```diff
@@ -227,7 +232,9 @@ - (instancetype)_initWithWebKitPropertyListData:(NSDictionary *)plist;
 
-    [dict setObject:m_data.url.toID().get() forKey:@"URL"];
+    auto nsURL = m_data.url.toID();
+    if (nsURL)
```
ditto.

cdumez commented on Source/WebKit/Shared/Cocoa/CoreIPCNSURLRequest.mm:
```diff
@@ -254,13 +261,18 @@ - (instancetype)_initWithWebKitPropertyListData:(NSDictionary *)plist;
-                    [headerFields setObject:(NSString *)s forKey:(NSString *)headerPair.first];
+                    auto key = (NSString *)headerPair.first;
+                    auto value = (NSString *)s;
+                    if (key && value)
```
This part of the change doesn't make sense. A null String() will give @"", not a nil `NSString*`.

cdumez replied:
If you want to skip those, you could check `isNull()` on the String directly, before converting to NSString*.

cdumez commented on Source/WebKit/Shared/Cocoa/CoreIPCNSURLRequest.mm:
```diff
@@ -254,13 +261,18 @@ - (instancetype)_initWithWebKitPropertyListData:(NSDictionary *)plist;
                     for (auto& item : vector)
                         [array addObject: item];
-                    [headerFields setObject:array.get() forKey:(NSString *)headerPair.first];
+                    auto key = (NSString *)headerPair.first;
```
ditto.

cdumez commented on Source/WebKit/Shared/Cocoa/CoreIPCNSURLRequest.mm:
```diff
@@ -274,7 +286,9 @@ - (instancetype)_initWithWebKitPropertyListData:(NSDictionary *)plist;
                 [&] (const auto& d) {
-                    [array addObject:d.toID().get()];
+                    auto obj = d.toID();
+                    if (obj)
```
Would reduce the scope of the variable.

cdumez commented on Source/WebKit/Shared/Cocoa/CoreIPCNSURLRequest.mm:
```diff
@@ -311,8 +325,11 @@ - (instancetype)_initWithWebKitPropertyListData:(NSDictionary *)plist;
-        for (auto& value : *m_data.contentDispositionEncodingFallbackArray)
-            [array addObject:value.toID().get()];
+        for (auto& value : *m_data.contentDispositionEncodingFallbackArray) {
+            auto valueObj = value.toID();
```
ditto.

cdumez commented on Tools/TestWebKitAPI/Tests/IPC/IPCSerialization.mm:
```diff
@@ -1335,6 +1335,33 @@ static void runTestCF(const CFHolderForTesting& holderArg)
+TEST(IPCSerialization, NSURLRequest)
+{
+    // NSURLRequest
+    NSURL *url = [NSURL URLWithString:@"https://webkit.org/"];
```
Please use RetainPtr

cdumez commented on Tools/TestWebKitAPI/Tests/IPC/IPCSerialization.mm:
```diff
@@ -1335,6 +1335,33 @@ static void runTestCF(const CFHolderForTesting& holderArg)
+{
+    // NSURLRequest
+    NSURL *url = [NSURL URLWithString:@"https://webkit.org/"];
+    NSURLRequest *urlRequest = [NSURLRequest requestWithURL:url];
```
ditto.

cdumez commented on Tools/TestWebKitAPI/Tests/IPC/IPCSerialization.mm:
```diff
@@ -1335,6 +1335,33 @@ static void runTestCF(const CFHolderForTesting& holderArg)
+        @"a" : @1,
+        @"b" : @2
+    };
+    NSData *postData = [NSJSONSerialization dataWithJSONObject:jsonDict options:0 error:Nil];
```
Ditto.

cdumez commented on Tools/TestWebKitAPI/Tests/IPC/IPCSerialization.mm:
```diff
@@ -1335,6 +1335,33 @@ static void runTestCF(const CFHolderForTesting& holderArg)
+        @"b" : @2
+    };
+    NSData *postData = [NSJSONSerialization dataWithJSONObject:jsonDict options:0 error:Nil];
+    NSMutableURLRequest *postRequest = [[NSMutableURLRequest alloc] initWithURL:url];
```
ditto.

sheeparegreat committed:
Update Serialization For CoreIPCNSURLRequest
https://bugs.webkit.org/show_bug.cgi?id=279092
rdar://134401345

Reviewed by Chris Dumez.

Add NULL checks and update tests to run on more platforms.

* Source/WebKit/Shared/Cocoa/CoreIPCNSURLRequest.mm:
(WebKit::CoreIPCNSURLRequest::toID const):
* Tools/TestWebKitAPI/Tests/IPC/IPCSerialization.mm:
(TEST(IPCSerialization, NSURLRequest)):
(TEST(IPCSerialization, SecureCoding)):

Canonical link: https://commits.webkit.org/283236@main
